Focus currently selected block when entering canvas

jeryj opened the pull request:
Now that the inserter stays open when focus leaves it, an edge case can occur where blocks are inserted before the canvas has ever been focused. When that happens, focus is trying to get sent to the previously focused block, but no block has been focused.

This focuses the currently selected block when entering the canvas.

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Have a post with some blocks 
- Reload the page
- Click the inserter (do not focus the editor canvas)
- Add a block
- Tab to the canvas
- The inserted block should be selected

jeryj committed:
Fix scenario where blocks have been inserted but focus has not left the canvas

Now that the inserter stays open when focus leaves it, an edge case can occur where blocks are inserted before the canvas has ever been focused. When that happens, focus is trying to get sent to the previously focused block, but no block has been focused.

This focuses the currently selected block when entering the canvas.

jeryj commented on packages/block-editor/src/components/writing-flow/use-tab-nav.js:
```diff
@@ -45,7 +45,11 @@ export default function useTabNav() {
 			container.current.focus();
 		} else if ( getSelectedBlockClientId() ) {
-			getLastFocus()?.current.focus();
+			container.current
+				.querySelector(
+					`[data-block="${ getSelectedBlockClientId() }"]`
+				)
+				.focus();
```
@ellatrix suggested this, so it's good :)

ellatrix replied:
We should probably add a non-hook util at some point similar to `useBlockRef`

mikachan reviewed (approved):
🚢

priethor reviewed (approved):
Works as expected ✅

jeryj commented:
@afercia, yes this is trying to preserve existing behavior that was broken by leaving the inserter open. Sorry for not being clearer about it. The discussion around keeping the inserter open is in https://github.com/WordPress/gutenberg/issues/61051.

jeryj committed:
Clear last focused block when inserting new blocks

When inserting a new block from the inserter, focus should go to this block when focus is moved to the canvas via the Tab keypress. To do this, we need to clear where focus previously was when a new block is inserted, otherwise it tries to place focus where it was before.

jeryj committed:
Only clear last focus if we're selecting the block or focusing it on insert

jeryj commented on packages/block-editor/src/components/inserter/hooks/use-insertion-point.js:
```diff
@@ -86,10 +87,23 @@ function useInsertionPoint( {
+				shouldFocusBlock ||
+				selectBlockOnInsert
+			) {
+				setLastFocus( null );
```
I considered doing this within the inserter sidebar menu, but I think it should live here as it will be an issue for anyone trying to insert new blocks from outside the editor canvas. This allows people to choose if they want focus to return to the canvas where it was (preserve last focus) or clear it and focus the new block.

jeryj commented on packages/block-editor/src/components/writing-flow/use-tab-nav.js:
```diff
@@ -45,7 +45,16 @@ export default function useTabNav() {
 			container.current.focus();
 		} else if ( getSelectedBlockClientId() ) {
-			getLastFocus()?.current.focus();
+			if ( getLastFocus()?.current ) {
+				getLastFocus().current.focus();
+			} else {
```
I had to preserve this, as focusing the selected block broke focusing inner blocks or things like focusing the table grid cell text field within a table block.

alexstine commented:
@jeryj I'm not sure how this was working before because someone got in a bit of a merge hurry but the experience is still quite bad.
1. Fresh pull of trunk and new build.
2. Open a post.
3. Select Toggle block inserter.
4. Select List to insert the block.
5. Notice how whole page has a focus loss.
6. Press Tab.
7. Focus lands back on the List block inside the inserter.
8. Notice how arrow keys no longer navigate the blocks to insert.
9. Press Tab to exit the inserter. Focus lands on block toolbar. I did not expect that to happen.
10. Next Tab finally places focus on the block.

Tested in Firefox.

Not sure if any of these bugs are related to this PR but this is really worrisome. One of the most used features of the editor, block inserter, is currently mostly unuseable for keyboard users.

Could we please refrain from merging to trunk before it can be checked for accessibility? I have no doubt this was probably an improvement over the experience but as mentioned above, these bugs are starting to come out of nowhere and it's really hard to identify which PR does it when there's never enough time for testing.

Thanks.

alexstine commented:
Also, another bug I found, it seems that when the inserter is open, pressing Escape in the canvas no longer activates navigation mode. 😞

jeryj commented:
Thanks for the testing @alexstine. I've got a PR for one and am looking at the other.

jeryj commented:
Switching modes should be fixed by https://github.com/WordPress/gutenberg/pull/61563.
